packages/cdk-assets/lib/private/handlers/files.ts

```diff
@@ -44,17 +43,42 @@ export class FileAssetHandler implements IAssetHandler {
       return;
     }
 
+    // Identify the the bucket encryption type to set the header on upload
+    // required for SCP rules denying uploads without encryption header
+    let paramsEncryption: {[index: string]:any}= {};
+    const encryption2 = await bucketEncryption(s3, destination.bucketName);
+    switch (encryption2) {
+      case BucketEncryption.NO_ENCRYPTION:
+        break;
+      case BucketEncryption.SSEAlgorithm_AES256:
+        paramsEncryption = { ServerSideEncryption: 'AES256' };
+        break;
+      case BucketEncryption.SSEAlgorithm_aws_kms:
+        paramsEncryption = { ServerSideEncryption: 'aws:kms' };
```

Comment:
Don't we potentially need to include the correct KMS key ARN? 

If the user has configured a custom key ARN as "DefaultEncryption" and we just pass `ServerSideEncryption: 'aws:kms'`, which key is going to be used? The user's key or `aws/s3` ?

Reply:
I tested the following setups
* encryption key type: a) Amazon S3 key (SSE-S3) -> yields to { ServerSideEncryption: 'AES256' };, no key configurable
* encryption key type: b) AWS Key Management Service key (SSE-KMS)
  * b1) AWS managed key (aws/s3)
  * b2) custom AWS KMS key

see details below for the configurations.

In summary:
We set the SSEAlgorithm, and the bucket automatically use whatever encryption is currently set up as default in the bucket. The bucket stores which is the default key to be used for default encryption. In fact the bucket would use the default encryption anyway but we need the header to get through the SCP rules.
I tried all the below scenarios below and deploy works.  It would not make sense for default encryption to use another key than configured in the bucket.

a)
```
{
    "ServerSideEncryptionConfiguration": {
        "Rules": [
            {
                "ApplyServerSideEncryptionByDefault": {
                    "SSEAlgorithm": "AES256"
                },
                "BucketKeyEnabled": false
            }
        ]
    }
}
```

b1)
```
{
    "ServerSideEncryptionConfiguration": {
        "Rules": [
            {
                "ApplyServerSideEncryptionByDefault": {
                    "SSEAlgorithm": "aws:kms",
                    "KMSMasterKeyID": "arn:aws:kms:us-east-1:703956386855:alias/aws/s3"
                },
                "BucketKeyEnabled": true
            }
        ]
    }
}
```

b2)
```
{
    "ServerSideEncryptionConfiguration": {
        "Rules": [
            {
                "ApplyServerSideEncryptionByDefault": {
                    "SSEAlgorithm": "aws:kms",
                    "KMSMasterKeyID": "arn:aws:kms:us-east-1:703956386855:key/833b292e-bc1f-47ec-af3f-1c1b6504815b"
                },
                "BucketKeyEnabled": true
            }
        ]
    }
}
```

![Screenshot 2021-11-24 at 15 13 44](https://user-images.githubusercontent.com/2978097/143257466-9af703d4-83af-49ee-9328-038fa0028be3.png)

Reply:
Hmm that's odd -- the behaviour we are seeing is that when CDK only includes the `x-amz-server-side-encryption: aws:kms` header, the objects end up being encrypted with `aws/s3` key, which currently is not permitted by our bucket policy.

Reply:
@vludax 
Do you have default bucket encryption enabled for the bucket? 
Did you use `cdk bootstrap` to reinitialize the cdk resources? 
Can you also provide the bucket policy, so that I can reproduce it?

Reply:
![image](https://user-images.githubusercontent.com/5164413/148311114-9fc0b43e-7bc6-4288-a50c-5bdc70392260.png)
The following bucket policy statement blocks the uploads -- when we remove this statement, the files get uploaded successfully and encrypted with `aws/s3`:
```
{
    "Sid": "Enforce default bucket encryption by rejecting any overrides",
    "Effect": "Deny",
    "Principal": "*",
    "Action": "s3:PutObject",
    "Resource": "arn:aws:s3:::<bucket-name>/*",
    "Condition": {
        "Null": {
            "s3:x-amz-server-side-encryption": "false"
        },
        "StringNotLikeIfExists": {
            "s3:x-amz-server-side-encryption-aws-kms-key-id": "arn:aws:kms:ap-southeast-2:XXX:key/YYY"
        }
    }
}
```

We did not run the bootstrap as we are using a custom synthesizer in our CDK -- that doesn't seem to specify encryption options anywhere. I'd also note that our uploads are working as expected with v1.118.0 of the CDK -- the files are successfully uploaded and encrypted with the right KMS key with the above policy statement in place.

Reply:
@vludax please note that this is already being addressed and fixed here: https://github.com/aws/aws-cdk/issues/18262